numpy/core/function_base.py

```diff
@@ -1,10 +1,24 @@
 from __future__ import division, absolute_import, print_function
 
+import warnings
+import operator
+
 __all__ = ['logspace', 'linspace']
 
 from . import numeric as _nx
 from .numeric import result_type, NaN, shares_memory, MAY_SHARE_BOUNDS, TooHardError
 
+def _index_deprecate(i, stacklevel=2):
```

Comment:
`__index__` is the standard Python protocol for "you should support silently coercing me to an integer", and what we use universally everywhere else (or have been migrating towards using). For the particular types you list either approach will give the same answer, but in general there's no rule that everyone who implements `__index__` has to also subclass `Integral`.

Also:
```
In [4]: isinstance(np.int8, numbers.Integral)
Out[4]: False
```